src/dwarf.cpp

```diff
@@ -304,137 +304,68 @@ void line_header::read(freader& s, const file_details& details) {
+    };
+    static const auto first = std::begin(skip_tags);
+    static const auto last = std::end(skip_tags);
 
-    d._type_resolved = true;
+    return std::find(first, last, d._tag) != last;
 }
 
 /**************************************************************************************************/
-#if 0
-bool resolve_specification_attribute(const dies& dies, die& d, const empool_callback& empool) { // REVISIT (fbrereto): d is an out-arg
-    // Section 2.13.2 (Declarations Completing Non-Defining Declarations) of the DWARF spec
-    // indicate that DIEs with a specification attribute do not need to duplicate the
-    // information present in the specification die. However, it _also_ says that not all
-    // attributes in the specification die will apply to _this_ die. So we have to pick
-    // and choose them. Nice. Here we make the expensive choice to copy over the attributes
-    // we care about from the specification die to this one. REVISIT (fosterbrereton) this
-    // is also sub-optimal, as it throws away the previous attribute memory block that gets
-    // allocated to make room for these other attributes.
-    if (!d.attribute_has_die(dw::at::specification)) return false;
-
-    const auto& spec_die = d.attribute_die(dw::at::specification);
-    const dw::at attributes_to_copy_over[] = {
-        dw::at::accessibility,
-        dw::at::linkage_name,
-    };
-
-    // replace with a std::accumulate when my head stops spinning from the problem at hand.
-    std::size_t additional{0};
-    for (const auto& attribute : attributes_to_copy_over) {
-        additional += spec_die.has_attribute(attribute);
-    }
-
-    if (additional == 0) return false;
 
-    std::size_t old_attr_count = d._attributes_size;
-    attribute* old_attr = d._attributes;
-
-    d._attributes_size += additional;
-    d._attributes = alloc_attributes(d._attributes_size);
-
-    attribute* next_attr = std::copy(old_attr, old_attr + old_attr_count, d._attributes);
-
-    for (const auto& attribute : attributes_to_copy_over) {
-        if (!spec_die.has_attribute(attribute)) continue;
-        *next_attr++ = spec_die.attribute(attribute);
-    }
+enum class process_mode {
+    complete,
+    single,
+};
 
-    return true;
-}
-#endif
 /**************************************************************************************************/
 
 } // namespace
 
 /**************************************************************************************************/
 
 struct dwarf::implementation {
-    implementation(object_ancestry&& ancestry,
-                   freader& s,
-                   const file_details& details,
-                   callbacks callbacks)
-        : _ancestry(std::move(ancestry)), _s(s), _details(details),
-          _callbacks(std::move(callbacks)) {}
+    implementation(const object_ancestry& ancestry,
+                   freader&& s,
+                   file_details&& details,
+                   register_dies_callback&& callback)
```

Comment:
Not wrong, but seems odd: why does a callback have move semantics?

Reply:
It's being sunk into the implementation, like the two parameters before it. Sink parameters are generally passed by rvalue-ref. The ancestry is passed by const-ref because we store a pointer to the instance, instead of (a copy of) the instance itself.